src/Internal/DependencyManagement/AbstractServiceProvider.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+ * AbstractServiceProvider class file.
+ *
+ * @package Automattic\WooCommerce\Internal\DependencyManagement
+ */
+
+namespace Automattic\WooCommerce\Internal\DependencyManagement;
+
+use League\Container\Argument\RawArgument;
+use League\Container\Definition\DefinitionInterface;
+use League\Container\Definition\Definition;
+
+/**
+ * Base class for the service providers used to register classes in the container.
+ *
+ * See the documentation of the original class this one is based on (https://container.thephpleague.com/3.x/service-providers)
+ * for basic usage details. What this class adds is:
+ *
+ * - The `add_with_auto_arguments` method that allows to register classes without having to specify the constructor arguments.
+ * - The `share_with_auto_arguments` method, sibling of the above.
+ * - Convenience `add` and `share` methods that are just proxies for the same methods in `$this->getContainer()`.
+ */
+abstract class AbstractServiceProvider extends \League\Container\ServiceProvider\AbstractServiceProvider {
+
+	/**
+	 * Register a class in the container and use reflection to guess the constructor arguments.
+	 *
+	 * WARNING: this method uses reflection, so please have performance in mind when using it.
+	 *
+	 * @param string $class_name Class name to register.
+	 * @param mixed  $concrete   The concrete to register. Can be a shared instance, a factory callback, or a class name.
+	 * @param bool   $shared Whether to register the class as shared (`get` always returns the same instance) or not.
+	 *
+	 * @return DefinitionInterface The generated container definition.
+	 *
+	 * @throws \Exception Error when reflecting the class, or class constructor is not public, or an argument has no valid type hint.
+	 */
+	protected function add_with_auto_arguments( string $class_name, $concrete = null, bool $shared = false ) : DefinitionInterface {
+		try {
+			$reflector = new \ReflectionClass( $class_name );
```

Comment:
We should actually take the `$concrete` argument into consideration here since it will be instantiated instead of `$class_name`. We could also support passing closures using `is_callable()` and `ReflectionFunction`!

Reply:
Something like:

```php
if ( ! isset( $concrete ) || is_string( $concrete ) && class_exists( $concrete ) ) {
			try {
				$class     = $concrete ?? $class_name;
				$reflector = new \ReflectionClass( $class );
				$reflector = $reflector->getConstructor();
				if ( isset( $reflector ) && ! $reflector->isPublic() ) {
					throw new \Exception( "AbstractServiceProvider::add_with_auto_arguments: constructor of class '$class_name' isn't public, instances can't be created." );
				}
			} catch ( \ReflectionException $ex ) {
				throw new \Exception( "AbstractServiceProvider::add_with_auto_arguments: error when reflecting class '$class': {$ex->getMessage()}" );
			}
		} elseif ( is_callable( $concrete ) ) {
			try {
				$reflector = new \ReflectionFunction( $concrete );
			} catch ( \ReflectionException $ex ) {
				throw new \Exception( "AbstractServiceProvider::add_with_auto_arguments: error when reflecting callable: {$ex->getMessage()}" );
			}
		} else {
			throw new \Exception( 'AbstractServiceProvider::add_with_auto_arguments: concrete must be a class or callable.' );
		}
```

Reply:
The original `Container` class already supports passing function names, closures and object instances as concretes, so we can support all of that for free and just add some validation. But yes, the constructor visibility must be checked when the concrete is a class name too.

Reply:
While the `Container` class does support passing `callable` it does not perform automatic argument resolution. By making use of `ReflectionFunction` we could do something like this:

```php
$factory = function ( DependencyA $dependency ) {
	return new Thing( $dependency );
};

$this->add_with_auto_args( Interface::class, $factory );
```

Then when you resolve that interface it will inject the dependency at execution time.

Reply:
The `add_with_auto_arguments` method is just a helper that creates the definition and invokes `addArgument` as appropriate, but what you are suggesting would involve extending the functionality of the core container itself which would be a somewhat complex task. I'm not saying it's a bad idea (quite the opposite), but I don't think we should delve into that right now; should we decide to add this functionality in the future, we will be able to do it without breaking anything.

Reply:
Also, we can always cover your example use case like this:

```
$factory = function() {
        $dependency = wc_get_container()->get( DependencyA::class );
	return new Thing( $dependency );
};
```

Not as beautiful but not terrible either.

Reply:
Since the underlying container does actually already support passing arguments to callables I've gone ahead and pushed a commit adding this feature as discussed!